// cleaner-wrasse/controllers/home.js
 var User = require('../../../models/User');
//set the plugin controller directory
 
 
///////////////////////////////////////////////
////     SET YOUR APP.JSON DETAILS        //// 
/////////////////////////////////////////////
//Not working ? try double dots on the json url..
var myModule = require('../../../app.json');
var sitename = myModule.sitename
var website = myModule.website
var description = myModule.description
var repo = myModule.repo

var wrasseModel = require('../models/wrasse.js'); 

///////////////////////////////////////
////       HOME CONTROLLER        //// 
/////////////////////////////////////
exports.index = function(req, res) {
   
  res.render('../../../public/homepage/index', {
        layout: false,
        siteName : sitename,
        description : description,    
        sitekey:process.env.SITE_KEY,
        getallissues:req.getallissues,
        openCount : req.openCount ,
        closedCount : req.closedCount,
        routing : res.locals.wrasserouting
  });
};
