// cleaner-wrasse/controllers/read.js
///////////////////////////////////////////////
////     SET YOUR APP.JSON DETAILS        //// 
/////////////////////////////////////////////
//Not working ? try double dots on the json url..
var myModule = require('../../../app.json');
var sitename = myModule.sitename
var website = myModule.website
var description = myModule.description
var repo = myModule.repo

var wrasseModel = require('../models/wrasse.js'); 

///////////////////////////////////////////////
///////   GET ALL ISSUES ALL USERS    ////////
/////////////////////////////////////////////
exports.getallissues = function(req, res, next) {
  console.log('Loading middleware , get all issues')
  //Determine how many forms exist on the server.
  var query1 = wrasseModel.find({ "parentid": { "$exists" : false } })
  
  query1.exec(function (err, docs) {
    if(err){console.log('Error Here'); return;}
    var openCount =0  
    for (var i = 0; i < docs.length; i++) {
      if (docs[i].active =='open') {
        openCount = openCount +1
      }
      var closedCount = docs.length-openCount//posible variation that dont work here.
    }
    req.getallissues = docs  
    req.openCount = openCount 
    req.closedCount = closedCount
   
    next();
  })


  
};

//////////////////////////////////
////       NEW ISSUE         //// 
////////////////////////////////
exports.newissue = function(req, res) {
  res.render('newissue', {
    siteName : sitename,
    description : description,
    sitekey:process.env.SITE_KEY,
    getallissues:req.getallissues,
    openCount : req.openCount ,
    closedCount : req.closedCount,
    routing : res.locals.wrasserouting
  });
};


//////////////////////////////////
////       NEW ISSUE GROUP   //// 
////////////////////////////////
exports.newissuegroup = function(req, res) {
  res.render('newissuegroup', {
    siteName : sitename,
    description : description,
    sitekey:process.env.SITE_KEY,
    getallissues:req.getallissues,
    openCount : req.openCount ,
    closedCount : req.closedCount,
    routing : res.locals.wrasserouting
  });
};

//////////////////////////////////
////       VIEW ISSUES        //// 
////////////////////////////////
exports.readissues = function(req, res) {
  res.render('viewissue', {
    siteName : sitename,
    description : description,
    sitekey:process.env.SITE_KEY,
    getallissues:req.getallissues,
    openCount : req.openCount ,
    closedCount : req.closedCount,
    routing : res.locals.wrasserouting
  });
};

//////////////////////////////////
////  VIEW ISSUES BY GROUP   //// 
////////////////////////////////
exports.readissuesbygroup = function(req, res) {
  res.render('viewissuebygroup', {
    siteName : sitename,
    description : description,
    sitekey:process.env.SITE_KEY,
    getallissues:req.getallissues,
    openCount : req.openCount ,
    closedCount : req.closedCount,
    routing : res.locals.wrasserouting
  });
};

//////////////////////////////////
////     VIEW ISSUE BY ID    //// 
////////////////////////////////
exports.readissuebyid = function(req, res) {
  res.render('viewissuebyid', {
    siteName : sitename,
    description : description,
    sitekey:process.env.SITE_KEY,
    wrassequery:req.wrassequery,
    wrassequeryparent:req.wrassequeryparent,
    routing : res.locals.wrasserouting
  });
};
////////////////////////////////////
////       UPDATE ISSUE        //// 
//////////////////////////////////
exports.updateissue = function(req, res) {
  res.render('updateissue', {
    siteName : sitename,
    description : description,
    sitekey:process.env.SITE_KEY,
    wrassequery:req.wrassequery,
    routing : res.locals.wrasserouting
  });
};

///////////////////////////////////////////////
///////   GET ONE ISSUES BY ID        ////////
/////////////////////////////////////////////
exports.getissue = function(req, res, next) {
 
  var receivedid =req.params['receivedid'];
  console.log('Loading middleware , get an issue : '+ receivedid)
  //Determine how many forms exist on the server.

  var wrassequery = wrasseModel.findOne({
                                  _id: receivedid}
                                  )
  wrassequery.exec(function (err, docs) {
    if(err){console.log('Error Here'); return;}
   
    req.wrassequery = docs  
    
   
    next();
  })
}
  
///////////////////////////////////////////////
///////   GET ISSUES BY PARENT ID     ////////
/////////////////////////////////////////////
exports.getparent = function(req, res, next) {
 
  var receivedid =req.params['receivedid'];
  console.log('Loading middleware , get an issue : '+ receivedid)
  //Determine how many forms exist on the server.

//
  var wrassequery = wrasseModel.find({
                                  parentid: receivedid}
                                  )
  wrassequery.exec(function (err, docs) {
    if(err){console.log('Error Here'); return;}
   
    req.wrassequeryparent = docs  
    
   
    next();
  })
}
